configtx/application.go

```diff
@@ -423,15 +431,359 @@ func (a *ApplicationOrg) setMSPConfig(updatedMSP MSP) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	certs := msp.TLSRootCerts[:]
+	for i, c := range msp.TLSRootCerts {
+		if c.Equal(cert) {
+			certs = append(certs[:i], certs[i+1:]...)
+			break
+		}
+	}
+
+	msp.TLSRootCerts = certs
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// AddTLSIntermediateCert adds an TLS intermediate cert trusted by the organization MSP.
+func (m *ApplicationOrgMSP) AddTLSIntermediateCert(cert *x509.Certificate) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	for _, c := range msp.TLSIntermediateCerts {
+		if c.Equal(cert) {
+			return nil
+		}
+	}
+
+	msp.TLSIntermediateCerts = append(msp.TLSIntermediateCerts, cert)
+
+	err = msp.validateCACerts()
+	if err != nil {
+		return err
+	}
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// RemoveTLSIntermediateCert removes a trusted TLS intermediate cert from the organization MSP.
+func (m *ApplicationOrgMSP) RemoveTLSIntermediateCert(cert *x509.Certificate) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	certs := msp.TLSIntermediateCerts[:]
+	for i, c := range msp.TLSIntermediateCerts {
+		if c.Equal(cert) {
+			certs = append(certs[:i], certs[i+1:]...)
+			break
+		}
+	}
+
+	msp.TLSIntermediateCerts = certs
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// SetClientOUIdentifier sets the NodeOUs client ou identifier for the organization MSP.
+func (m *ApplicationOrgMSP) SetClientOUIdentifier(clientOU membership.OUIdentifier) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	msp.NodeOUs.ClientOUIdentifier = clientOU
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// SetPeerOUIdentifier sets the NodeOUs peer ou identifier for the organization MSP.
+func (m *ApplicationOrgMSP) SetPeerOUIdentifier(peerOU membership.OUIdentifier) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	msp.NodeOUs.PeerOUIdentifier = peerOU
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// SetAdminOUIdentifier sets the NodeOUs admin ou identifier for the organization MSP.
+func (m *ApplicationOrgMSP) SetAdminOUIdentifier(adminOU membership.OUIdentifier) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	msp.NodeOUs.AdminOUIdentifier = adminOU
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// SetOrdererOUIdentifier sets the NodeOUs orderer ou identifier for the organization MSP.
+func (m *ApplicationOrgMSP) SetOrdererOUIdentifier(ordererOU membership.OUIdentifier) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	msp.NodeOUs.OrdererOUIdentifier = ordererOU
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// SetEnableNodeOUs sets the NodeOUs recognition, if NodeOUs recognition is enabled then an msp identity
+// that does not contain any of the specified OUs will be considered invalid.
```

Comment:
`that does not contain exactly one of the fabric Node OU identifiers will be considered invalid.`

Reply:
🛠️